modules/core/src/main/java/org/apache/fluo/core/impl/TransactionImpl.java

```diff
@@ -872,447 +826,688 @@ public int getSize() {
-              DelLockValue.encodeRollback(false, true));
+    return firstStep.compose(cd)
+        .thenRun(() -> cd.commitObserver.commitFailed(cd.getShortCollisionMessage()));
+
+  }
+
+
+  class RollbackOtherLocks extends BatchWriterStep {
+
+    @Override
+    public Collection<Mutation> createMutations(CommitData cd) {
+      // roll back locks
+
+      // TODO let rollback be done lazily? this makes GC more difficult
+
+      Flutation m;
+
+      ArrayList<Mutation> mutations = new ArrayList<>(cd.acceptedRows.size());
+      for (Bytes row : cd.acceptedRows) {
+        m = new Flutation(env, row);
+        for (Entry<Column, Bytes> entry : updates.get(row).entrySet()) {
+          if (isReadLock(entry.getValue())) {
+            m.put(entry.getKey(),
+                ColumnConstants.RLOCK_PREFIX | ReadLockUtil.encodeTs(startTs, true),
+                DelReadLockValue.encodeRollback());
+          } else {
+            m.put(entry.getKey(), ColumnConstants.DEL_LOCK_PREFIX | startTs,
+                DelLockValue.encodeRollback(false, true));
+          }
         }
+        mutations.add(m);
       }
-      mutations.add(m);
+
+      return mutations;
     }
+  }
+
+  class RollbackPrimaryLock extends BatchWriterStep {
+
+    @Override
+    public Collection<Mutation> createMutations(CommitData cd) {
+      // mark transaction as complete for garbage collection purposes
+      Flutation m = new Flutation(env, cd.prow);
+
+      m.put(cd.pcol, ColumnConstants.DEL_LOCK_PREFIX | startTs,
+          DelLockValue.encodeRollback(startTs, true, true));
+      m.put(cd.pcol, ColumnConstants.TX_DONE_PREFIX | startTs, EMPTY);
 
-    CompletableFuture<Void> cfuture =
-        env.getSharedResources().getBatchWriter().writeMutationsAsyncFuture(mutations);
-    addCallback(cfuture, cd, result -> rollbackPrimaryLock(cd));
+      return Collections.singletonList(m);
+    }
   }
 
-  private void rollbackPrimaryLock(CommitData cd) throws Exception {
+  @VisibleForTesting
+  public boolean commitPrimaryColumn(CommitData cd, Stamp commitStamp) {
 
-    // mark transaction as complete for garbage collection purposes
-    Flutation m = new Flutation(env, cd.prow);
+    SyncCommitObserver sco = new SyncCommitObserver();
+    cd.commitObserver = sco;
+    try {
+      CommitStep firstStep = new GetCommitStampStepTest(commitStamp);
 
-    m.put(cd.pcol, ColumnConstants.DEL_LOCK_PREFIX | startTs,
-        DelLockValue.encodeRollback(startTs, true, true));
-    m.put(cd.pcol, ColumnConstants.TX_DONE_PREFIX | startTs, EMPTY);
+      firstStep.andThen(new WriteNotificationsStep()).andThen(new CommitPrimaryStep());
 
-    CompletableFuture<Void> cfuture =
-        env.getSharedResources().getBatchWriter().writeMutationsAsyncFuture(m);
-    addCallback(cfuture, cd,
-        result -> cd.commitObserver.commitFailed(cd.getShortCollisionMessage()));
+      firstStep.compose(cd).thenRun(() -> cd.commitObserver.committed())
```

Comment:
@jkosh44 I tracked down why the test are failing.  Its because `() -> cd.commitObserver.committed()` is always run.  This code is run even when `CommitPrimaryStep()` fails and calls `commitFailed()` on `SyncCommitObserver`.  So its a timing issue, will `() -> cd.commitObserver.committed()` or code in `sco.waitForCommit()` run first.

One way to resolve this would be to create simple commit step that calls `cd.commitObserver.committed()`.  Since its a commit step, it will only run if `CommitPrimaryStep()` is successful.

I tracked this down by adding code to `SyncCommitObserver` to print stack traces whenever any method on it is called.  Need to open a new issue to ensure only one method can be called on `SyncCommitObserver`.

Reply:
I opened #1011

Reply:
Awesome, the tests are all passing now.  I think this PR is done, I am going to give the diffs one more look today before merging.